### DISCUSSION TITLE
Use of Android OS SQLite library

### DISCUSSION THREAD

**Discussion by pourqavam** [simolus3/drift#1626] [Q&A] [answered]
Hi
First thanks for the Drift package.
1. How can I use the native (OS) SQLite library on Android, not the one provided by `sqlite3_flutter_libs`? (I don't need the new features of the SQLite)
I got this error:
```
Invalid argument(s): Failed to load dynamic library '/data/data/com.pack.../lib/libsqlite3.so': dlopen failed: libr...
#0      _open (dart:ffi-patch/ffi_dynamic_library_patch.dart:12:43)
#1      new DynamicLibrary.open (dart:ffi-patch/ffi_dynamic_library_patch.dart:23:12)
#2      _defaultOpen (package:sqlite3/src/load_library.dart:40:31)
#3      OpenDynamicLibrary.openSqlite (package:sqlite3/src/load_library.dart:122:12)
#4      sqlite3 (package:sqlite3/src/api/sqlite3.dart:10:38)
#5      _VmDelegate._createDatabase (package:drift/src/ffi/database.dart:169:13)
#6      _VmDelegate.open (package:drift/src/ffi/database.dart:148:7)
#7      DelegatedDatabase.ensureOpen.<anonymous closure> (package:drift/src/runtime/executor/helpers/engines.dart:301:22)
<asynchronous suspension>
#8      DatabaseConnectionUser.transaction (package:drift/src/runtime/api/connection_user.dart:416:12)
<asynchronous suspension>
...
```
2. Is there any performance benefits to use the OS library?
3. Will the two problems mentioned in the readme for the `sqlite3_flutter_libs`, still exist if we use the OS lib?

Thanks

My code:
```dart
@DriftDatabase(
  tables: [..........],
)
class Database extends _$Database {
  Database() : super(_openConnection());

  @override
  int get schemaVersion => _version;

  @override
  MigrationStrategy get migration => MigrationStrategy(
        beforeOpen: (details) async {
          await customStatement('PRAGMA foreign_keys = ON');
        },
      );
}

LazyDatabase _openConnection() {
  return LazyDatabase(() async {
    final dbFolder = await getApplicationDocumentsDirectory();
    final file = File(path.join(dbFolder.path, _fileName));
    return NativeDatabase(file, logStatements: kDebugMode);
  });
}
```

My dependencies:
```yaml
...
path_provider: ^2.0.8
path: ^1.8.0
drift: ^1.2.0
...
drift_dev: ^1.2.0
build_runner: ^2.1.7
```

**Comment by simolus3** [comment, marked as answer]
> I got this error:

You can't use `dart:ffi` (or a `NativeDatabase`) to access the sqlite3 library that comes with the operating system (well, at least not [since Android 7](https://ericsink.com/entries/sqlite_android_n.html)). It's possible on iOS, but on Android you'd have to go through the Java APIs. Even when using drift, you can use the [moor_flutter](https://pub.dev/packages/moor_flutter) package to use `sqflite` as a drift backend. `sqflite` uses the Java APIs internally.

> Is there any performance benefits to use the OS library?

I don't see why there should be a performance benefit. You binary size will certainly reduce (by around ~700 KiB when using app bundles). But when accessing the OS library through platform channels (which is necessary on Android), the overhead of serialization makes things much slower. This doesn't really matter for most apps, but a `NativeDatabase` is generally much faster than the `FlutterQueryExecutor` from `moor_flutter`.

> Will the two problems mentioned in the readme for the `sqlite3_flutter_libs`, still exist if we use the OS lib?

The first one is not a concern anymore (since we can't load the library shipped with the OS at all), the second should also be fine (IIRC Android configures sqlite3 to take care of that automatically).
